report.typ

```diff
@@ -0,0 +1,1689 @@
+bad wines. Make sure to use the same best-practices (train/test split,
+cross-validation) as we did before. Note that the regularization of the
+GBTs is a lot more complex than for Logistic Regression. Test the
+following parameters:
+
+```py
+param_grid = {
+  "max_depth": [3, 4, 5],  # Focus on shallow trees to reduce complexity
+  "learning_rate": [0.01, 0.05, 0.1],  # Slower learning rates
+  "n_estimators": [50, 100],  # More trees but keep it reasonable
+  "min_child_weight": [1, 3],  # Regularization to control split thresholds
+  "subsample": [0.7, 0.9],  # Sampling rate for boosting
+  "colsample_bytree": [0.7, 1.0],  # Sampling rate for columns
+  "gamma": [0, 0.1],  # Regularization to penalize complex trees
+}
+```
+
+]
+#block[
+```python
+from sklearn.preprocessing import RobustScaler
+
+
+param_grid = {
+    "max_depth": [3, 4, 5],  # Focus on shallow trees to reduce complexity
+    "learning_rate": [0.01, 0.05, 0.1],  # Slower learning rates
+    "n_estimators": [50, 100],  # More trees but keep it reasonable
+    "min_child_weight": [1, 3],  # Regularization to control split thresholds
+    "subsample": [0.7, 0.9],  # Sampling rate for boosting
+    "colsample_bytree": [0.7, 1.0],  # Sampling rate for columns
+    "gamma": [0, 0.1],  # Regularization to penalize complex trees
+}
+
+scaler = RobustScaler()
+
+X_scaled = scaler.fit_transform(X[numerical_features])
+X_scaled = pd.DataFrame(X_scaled, columns=X[numerical_features].columns)
+
+xgb_model = xgb.XGBClassifier(objective="binary:logistic")
+
+grid = GridSearchCV(xgb_model, param_grid, cv=10, n_jobs=-1, scoring="f1")
+
+f = X.columns.drop("color")
+X_red = X_scaled[X["color"] == "red"]
+y_red = df_binary_quality[X["color"] == "red"]
+
+X_train, X_test, y_train, y_test = train_test_split(
+    X_red, y_red, test_size=0.2, stratify=y_red
+)
+
+grid.fit(X_train, y_train)
+
+print("Best hyperparameters: ", grid.best_params_)
+print("Best score: ", grid.best_score_)
+print("Test score: ", grid.score(X_test, y_test))
+```
+
+#block[
+```
+Best hyperparameters:  {'colsample_bytree': 0.7, 'gamma': 0.1, 'learning_rate': 0.1, 'max_depth': 5, 'min_child_weight': 1, 'n_estimators': 100, 'subsample': 0.7}
+Best score:  0.6418544653027412
+Test score:  0.56
+```
+
+]
+]
+#block[
+- Analyze the results (test and train), validate whether there is
+  overfitting.
+
+]
+#block[
+```python
+train_sizes, train_scores, test_scores = learning_curve(
+    grid.best_estimator_,
+    X_train,
+    y_train,
+    cv=ShuffleSplit(n_splits=20, test_size=0.2),
+    scoring="f1",
+)
+
+display = LearningCurveDisplay(
+    train_sizes=train_sizes,
+    train_scores=train_scores,
+    test_scores=test_scores,
+    score_name="f1-score",
+)
+display.plot()
+plt.show()
+```
+
+#block[
+#box(image("--template=isc_templace.typ/163717f4158014aae9be6b12b6481f2984fbc890.png"))
+
+]
+]
+#block[
+OK nice overfitting, very good. That\'s what we like to see.
```

Comment:
I am extremely surprised to read that. Overfitting is a problem, not something we want to see, on the contrary.